src/GitHub.TeamFoundation.14/Services/VSGitExt.cs

```diff
@@ -1,39 +1,114 @@
 using System;
+using System.Linq;
 using System.Collections.Generic;
 using System.ComponentModel.Composition;
-using System.Linq;
-using GitHub.Extensions;
 using GitHub.Models;
 using GitHub.Services;
+using GitHub.Logging;
+using Serilog;
 using Microsoft.VisualStudio.TeamFoundation.Git.Extensibility;
+using Task = System.Threading.Tasks.Task;
 
 namespace GitHub.VisualStudio.Base
 {
+    /// <summary>
+    /// This service acts as an always available version of <see cref="IGitExt"/>.
+    /// </summary>
     [Export(typeof(IVSGitExt))]
     [PartCreationPolicy(CreationPolicy.Shared)]
     public class VSGitExt : IVSGitExt
     {
+        static readonly ILogger log = LogManager.ForContext<VSGitExt>();
+
+        IGitHubServiceProvider serviceProvider;
+        IVSUIContext context;
         IGitExt gitService;
 
-        public void Refresh(IServiceProvider serviceProvider)
+        [ImportingConstructor]
+        public VSGitExt(IGitHubServiceProvider serviceProvider, IVSUIContextFactory factory)
         {
-            if (gitService != null)
-                gitService.PropertyChanged -= CheckAndUpdate;
-            gitService = serviceProvider.GetServiceSafe<IGitExt>();
-            if (gitService != null)
-                gitService.PropertyChanged += CheckAndUpdate;
+            this.serviceProvider = serviceProvider;
+
+            // The IGitExt service is only available when in the SccProvider context.
+            // This could be changed to VSConstants.UICONTEXT.SolutionExists_guid when testing.
+            context = factory.GetUIContext(new Guid(Guids.GitSccProviderId));
+
+            // Start with empty array until we have a change to initialize.
+            ActiveRepositories = new ILocalRepositoryModel[0];
+
+            // If we're not in the GitSccProvider context or TryInitialize fails, have another go when the context changes.
+            if (context.IsActive && TryInitialize())
+            {
+                //// RefreshActiveRepositories on background thread so we don't impact startup.
+                InitializeTask = Task.Run(() => RefreshActiveRepositories());
+            }
+            else
+            {
+                context.UIContextChanged += ContextChanged;
+                log.Information("VSGitExt will be initialized later");
+                InitializeTask = Task.CompletedTask;
+            }
+        }
+
+        void ContextChanged(object sender, VSUIContextChangedEventArgs e)
+        {
+            // If we're in the GitSccProvider context and TryInitialize succeeds, we can stop listening for events.
```

Comment:
We need to continually listen for events so we stop using GitExt if the scc provider becomes deactivated.

Reply:
Could you explain why this is? Isn't @jcansdale's analysis correct? Looking at the code in ILSpy I don't see how `IGitExt` can become null once we have an instance unless the package is unloaded (which I don't think can happen?).